### ISSUE TITLE
RFC: Adopt Relay Hooks in favor of container-based APIs for managing GraphQL data

### ISSUE THREAD

**Issue by olerichter00** [artsy/eigen#6080]
## Proposal

Adopt [Relay Hooks](https://relay.dev/docs/migration-and-compatibility/) in favor of the container-based APIs for fetching and managing GraphQL data in Eigen ([Relay Hooks Migration Guide](https://relay.dev/docs/migration-and-compatibility/relay-hooks-and-legacy-container-apis/)). 

## Reasoning

1. Using Relay Hooks will require less code and has less indirections than the equivalent container-based solution.
2. The Hooks API gives us the ability to configure more granular fetch policies and comes with an improved TypeScript coverage (according to the [Relay documentation](https://relay.dev/blog/2021/03/09/introducing-relay-hooks/)).
4. By extracting queries it will be easy to import them in other places of the app which can for example be used for our query prefetching implementation.
5. The Relay team suggests using the hooks API and has removed examples using the old container-based APIs in the documentation. 
6. Using Relay Hooks will make it easier for engineers to rely on the official documentation.
7. The new `usePreloadedQuery` hook can be used to implement the "render-as-you-fetch" pattern to make data available faster.

## Exceptions

The old container-based APIs are still fully supported in all new Relay versions to this point and even can be mixed with the new Relay hooks API. This means we can incrementally adopt the new API and don't need to convert already existing components.

The old container-based APIs are still fully supported in all released Relay versions and can even be mixed with the new Relay Hooks API. This means that we can introduce the new API gradually and do not have to convert already existing components.

## Additional Context

After a [first attempt](https://github.com/artsy/eigen/issues/3562) to use Relay hooks with the [relay-hooks library](https://github.com/relay-tools/relay-hooks) which was not fully compatible with Relay's QueryRenderer, we realized that it was not possible to combine the container-based APIs from Relay and the hooks from the unofficial [relay-hooks library](https://github.com/relay-tools/relay-hooks) without using nested query renders. With the new official Relay hooks API, this is no longer an issue.

As soon as other clients upgrade Relay to a version that supports Relay hooks, this RFC can be applied to these projects as well.

["Introducing Relay Hooks" Blog Post](https://relay.dev/blog/2021/03/09/introducing-relay-hooks/)
## How is this RFC resolved?

We'll consider this RFC approved if 50% of the engineering team actively approves of this change or there are no major objections that require. Add a 👍 or 👎 reaction to this proposal to vote.

## Next Steps

In case of approval, we will start updating all Relay-related documentation and encourage engineers to use the new Relay hooks API for all new components in Eigen. 

[Relay Hooks Migration Guide](https://relay.dev/docs/migration-and-compatibility/relay-hooks-and-legacy-container-apis/)

Co-authored by @MounirDhahri

**Comment by mzikherman** [comment]
I don't code much in Eigen, but I had a general question about this.

I was on a team working with React/Relay on the web 2+ years ago, where we were using React classes and containers (I assume you mean `createFragmentContainer` and the like). The containers seemed like a nice abstraction (wrapper around React classes), and I enjoyed using React classes (coming from an object-oriented background).

Now, it seems like functional components and hooks are all the rage (while classes + containers remain first-class supported citizens as well). I haven't worked much in web since then, so I'm behind, but I sort of missed the memo/blog post on _why_ this new approach is better. I'm not sure why I need to essentially 'discard' some of that ecosystem knowledge, and learn something new in favor of hooks and functional components. The reasoning given above doesn't feel super compelling to me (although granted I am out of date and it seems like this ecosystem is pretty cutting-edge and changing often).

Can someone help explain (or post explanatory links) detailing this shift?

**Comment by olerichter00** [comment]
This is a very good question. Maybe [this chapter](https://reactjs.org/docs/hooks-intro.html#motivation) from the React docs can explain what problems functional components and hooks solve.

[This post](https://relay.dev/blog/2021/03/09/introducing-relay-hooks/) from the Relay blog gives an overview of the advantages of the new Relay Hooks API and explains why the Relay team is moving in this direction.

**Comment by MounirDhahri** [comment]
> I was on a team working with React/Relay on the web 2+ years ago, where we were using React classes and containers (I assume you mean createFragmentContainer and the like). The containers seemed like a nice abstraction (wrapper around React classes), and I enjoyed using React classes (coming from an object-oriented background).

This is a good point, we actually brought this into discussion a few months ago on [slack](https://artsy.slack.com/archives/C02BAQ5K7/p1599676269030100) (sorry non-artsy engineers readers), and we had an agreement that we liked having that abstraction serving us for the Separation of Concerns between the fetching logic and the UI logic. This same separation was not easy to achieve back with the unofficial relay hooks and we decided to stick with our containers-based approach. The situation is different now since we can achieve that using relay cooks and we can still keep a pattern that our engineers are familiar with and have a clear separation of concerns to the same granular level as before or further if we want to. 
This is what an Artist screen looks like in our current approach:
![old](https://user-images.githubusercontent.com/11945712/150828231-3c4d245d-a513-432a-9b66-efd7ca66f5c9.png)

This is what an Artist screen would look like in the suggested approach
![new](https://user-images.githubusercontent.com/11945712/150828766-165143b3-a1d7-4170-98e7-86261f5f6f04.png)


___

> I'm not sure why I need to essentially 'discard' some of that ecosystem knowledge, and learn something new in favor of hooks and functional components.

I think this RFC will not to discard any of the ecosystem knowledge our engineers acquired.  Since we stopped having any class components inside Eigen (and also other front-end clients I believe) for the last 1.5 years, we think it won't take much time for our engineers to get familiar with relay hooks instead of the HOCs approach since they're already familiar with the approach and the react docs as well are encouraging the usage of Hooks. This migration though will help our engineers be able to get information from the official relay docs which are now almost entirely written using hooks
